### PR TITLE
Fix crypto and Implement recover

### PR DESCRIPTION
**Pull request by chemonoworld**
1. Change signature to recoverable signature
2. Implement  recover in Signature

### CODE REVIEW

File: common/src/crypto.rs [hunk 1]
```diff
@@ -306,8 +335,8 @@ pub fn generate_keypair(seed: impl AsRef<[u8]>) -> (PublicKey, PrivateKey) {
     let secp = Secp256k1::new();
     let (private_key, public_key) = secp.generate_keypair(&mut rng);
     (
-        PublicKey::from_array(public_key.serialize()).expect("invalid public key"),
-        PrivateKey::from_array(private_key.secret_bytes()).expect("invalid private key"),
+        PublicKey::from_array(public_key.serialize()).expect("public key is invalid"),
+        PrivateKey::from_array(private_key.secret_bytes()).expect("private key is invalid"),
```
Comment (junha1) [inline common/src/crypto.rs:345]: Why did you change this?

File: common/src/crypto.rs [hunk 2]
```diff
@@ -115,43 +118,71 @@ impl fmt::Display for Hash256 {
...
     }
 
     /// Creates a new signature from the given data and keys.
     pub fn sign(data: Hash256, private_key: &PrivateKey) -> Result<Self, Error> {
         let private_key = secp256k1::SecretKey::from_slice(&private_key.key.data)
             .map_err(|_| Error::InvalidFormat("private key: [omitted]".to_owned()))?;
-        let secp = Secp256k1::new();
-        let message = Message::from_slice(data.as_ref()).unwrap();
+        let message = Message::from_slice(data.as_ref())
+            .map_err(|_| Error::InvalidFormat("message: data is invalid format".to_owned()))?;
```
Comment (junha1) [inline common/src/crypto.rs:136]: When does this fail?

File: common/src/crypto.rs [hunk 3]
```diff
@@ -115,43 +118,71 @@ impl fmt::Display for Hash256 {
...
+        let (recovery_id, rs) = Secp256k1::signing_only()
+            .sign_ecdsa_recoverable(&message, &private_key)
+            .serialize_compact();
+        let v = recovery_id.to_i32() as u8;
+        let bytes: [u8; 65] = {
+            let mut whole: [u8; 65] = [0; 65];
+            let eth_offset = 27;
+            let (left, right) = whole.split_at_mut(rs.len());
```
Comment (junha1) [inline common/src/crypto.rs:144]: is `rs.len()` dynamic?

File: common/src/crypto.rs [hunk 4]
```diff
@@ -280,11 +311,9 @@ impl PrivateKey {
     }
 
     pub fn public_key(&self) -> PublicKey {
-        let private_key =
-            secp256k1::SecretKey::from_slice(&self.key.data).expect("invalid private key");
-        let secp = Secp256k1::new();
-        let public_key = private_key.public_key(&secp);
-        PublicKey::from_array(public_key.serialize()).expect("invalid public key")
+        let private_key = SecretKey::from_slice(&self.key.data).expect("private key is invalid");
```
Comment (junha1) [inline common/src/crypto.rs:320]: ditto

File: common/src/crypto.rs (hunk 3 above)
Comment (chemonoworld) [reply common/src/crypto.rs:144]: No, fixed(64)
